### PR TITLE
Initial implementation of Task feature in CMS Core configuration code

### PR DESCRIPTION
**Pull request by wddgit**
This includes the initial implementation of the Task feature in the Core CMS configuration code. It allows creating Tasks, adding modules to them, and associating them to sequences, paths, end paths, and the schedule. In many ways they behave like Sequences except that the order modules appear on Tasks does not matter. New unit tests were added to test the Core code changes.

This PR also includes fixes for the runTheMatrix.py tests that I could run on cmsdev04 and IB unit tests. The intent of these python fixes was to have exactly the same modules run in unscheduled mode as before this change so the behavior should not change at all. I ran tests to verify this. But in the long run it would be a positive development for experts to use the Task feature to specify a smaller set of modules so resources are not wasted creating modules that will never need to be run. (I will monitor the full IBs and centrally fix any additional failures that show up)

There is much work remaining to be done to migrate all CMS configurations to use the Task feature.
This PR only does the minimum amount of migration needed to keep runTheMatrix.py and the unit tests working. A large amount of additional work will be needed to convert all CMS configuration code to use this new Task feature to specify modules that should be run in unscheduled mode instead of relying on the convertToUnscheduled function.

This PR modifies the convertToUnscheduled function to work properly with the new Task feature. EDProducers and EDFilters removed from paths and endpaths are automatically placed on a Task.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by slava77** [comment]
@wddgit 
could you please confirm that the same modules are running  per event processing.
Also, what about the modules constructed but not running.

**Comment by wddgit** [comment]
All the unit test and RelVal failures are unrelated to this PR. Details below:

The 3 RelVal errors are of category "DatabaseError" and occurred in the corresponding nightly IB tests without the changes here.

The unit test failure in RecoLocalCalo/HGCalRecProducers is also already occurring in the nightly IB tests.

The PythonParameterSet unit test fails when the packages Geometry/TrackerSimData and FWCore/PythonParameterSet are both checked out together in the same working area. This is a failure of the test that existed before my changes. It is related to the difficulty of testing a particular FileInPath behavior. If I remember correctly, Bill was trying to fix this before, but it is tricky testing this behavior of FileInPath in the unit testing environment and I guess he did not succeed in fixing this corner case. It is rare that both of those packages are checked out in the same working area and it does not occur in the full build IB case.

**Comment by Dr15Jones** [comment]
please test

**Comment by wddgit** [comment]
@slava77 Hi Slava, In response to your comment I ran the following additional test. I ran step3 of runTheMatrix test 50202.0 with the Tracer added. I did this in an unmodified working area and also repeated it in a working area with the changes in this PR. Then I ran grep to select all the lines about constructing modules and sent that to a file in both places. The results were identical. Then I did the same with the lines about a module processing an event. The results were also identical. They were even in the same order.

Before submitting the PR, I had run all the runTheMatrix tests and printed out a list of the modules constructed in unscheduled mode and done a similar comparison for all steps of all the tests I could run. That was also OK. That was the part the changes I made were most likely to break.

Is that sufficient? Let me know if there is something more you would like me to check.

**Comment by slava77** [comment]
On 11/10/16 9:11 PM, W. David Dagenhart wrote:
> @slava77 <https://github.com/slava77> Hi Slava, In response to your
> comment I ran the following additional test. I ran step3 of runTheMatrix
> test 50202.0 with the Tracer added. I did this in an unmodified working
> area and also repeated it in a working area with the changes in this PR.
> Then I ran grep to select all the lines about constructing modules and
> sent that to a file in both places. The results were identical. Then I
> did the same with the lines about a module processing an event. The
> results were also identical. They were even in the same order.
> 
> Before submitting the PR, I had run all the runTheMatrix tests and
> printed out a list of the modules constructed in unscheduled mode and
> done a similar comparison for all steps of all the tests I could run.
> That was also OK. That was the part the changes I made were most likely
> to break.
> 
> Is that sufficient? Let me know if there is something more you would
> like me to check.


I think this is sufficient to understand the impact of the PR.
I still need to look at the code changes

**Comment by wddgit** [comment]
The unit test failures in the second run of the test are unrelated to this PR. Two of them are the same as in the first run of the tests, see the my previous comments above about those.

A new one showed up this time that was not there the first time related to a file open. This failure has shown up in previous nightly IB tests so it cannot be caused by my changes. It is sporadic and only fails sometimes. I do not understand how it can happen. The test runs a shell script and the first step generates an output file and seems to succeed. Then a subsequent cmsRun process tries to read that file and fails to be able to open it. I cannot see any way this could be related to this PR. For example, here is a link to one of the failures in the nightly IB.

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc6_amd64_gcc530/CMSSW_8_1_X_2016-11-09-1100/unitTestLogs/RecoMET/METProducers

File: FWCore/Framework/src/StreamSchedule.cc [hunk 1]
```diff
@@ -207,16 +205,11 @@ namespace edm {
       // 2) if it is a WorkerT<EDProducer>, add it to our list
       // 3) hand list to our delayed reader
       for (auto const& label : unusedLabels) {
```
Comment (Dr15Jones) [inline FWCore/Framework/src/StreamSchedule.cc]: At this point, is `unusedLabels` actually an accurate variable name? Maybe `labelsForUnscheduled`?

File: FWCore/Framework/test/test_offPath_allowUnscheduled_false_fail_cfg.py [hunk 2]
```diff
@@ -4,7 +4,6 @@
 
 import FWCore.Framework.test.cmsExceptionsFatalOption_cff
 process.options = cms.untracked.PSet(
-    allowUnscheduled = cms.untracked.bool(False),
```
Comment (Dr15Jones) [inline FWCore/Framework/test/test_offPath_allowUnscheduled_false_fail_cfg.py]: Seems the `allowUnscheduled_*` file names are not really appropriate anymore

File: FWCore/ParameterSet/python/Config.py [hunk 3]
```diff
@@ -361,56 +386,89 @@ def __setattr__(self,name,value):
...
+                      raise ValueError(msg1+"sequence "+s.label_()+msg2)
+                   s = self.__findFirstUsingModule(self.paths,oldValue)
                    if s is not None:
-                      raise ValueError(msg1+s.label_()+msg2)
-                   s = self.__findFirstSequenceUsingModule(self.endpaths,oldValue)
+                      raise ValueError(msg1+"path "+s.label_()+msg2)
+                   s = self.__findFirstUsingModule(self.endpaths,oldValue)
                    if s is not None:
-                      raise ValueError(msg1+s.label_()+msg2)
-            self.__delattr__(name)
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: This function name has special meaning to python, why did you change it?

File: FWCore/ParameterSet/python/Config.py [hunk 4]
```diff
@@ -513,6 +574,8 @@ def addSubProcess(self,mod):
         self.__subProcesses.append(mod)
     def _placeService(self,typeName,mod):
         self._place(typeName, mod, self.__services)
+        if typeName in self.__dict__:
+          self.__dict__[typeName]._inProcess = False
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: What is this logic doing?

File: FWCore/ParameterSet/python/Config.py [hunk 5]
```diff
@@ -707,6 +785,40 @@ def _sequencesInDependencyOrder(self):
             raise RuntimeError("circular sequence dependency discovered \n"+
                                ",".join([label for label,junk in dependencies.iteritems()]))
         return returnValue
+    def _tasksInDependencyOrder(self):
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: Given the similarities between Tasks and Sequences in this case, is it feasible for the two to use the same code to do this check?

File: FWCore/ParameterSet/python/Config.py [hunk 6]
```diff
@@ -727,18 +839,17 @@ def dumpPython(self, options=PrintOptions()):
...
+        result+=self._dumpPythonList(self._sequencesInDependencyOrder(), options)
+        result+=self._dumpPythonList(self.paths_(), options)
+        result+=self._dumpPythonList(self.endpaths_(), options)
         result+=self._dumpPythonList(self.aliases_(), options)
-        if self.schedule:
-            pathNames = ['process.'+p.label_() for p in self.schedule]
-            result +='process.schedule = cms.Schedule(*[ ' + ', '.join(pathNames) + ' ])\n'
-
+        if not self.schedule_() == None:
+            result += 'process.schedule = ' + self.schedule.dumpPython(options)
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: Are you sure this works if there are more than 255 paths?

File: FWCore/ParameterSet/python/Config.py [hunk 7]
```diff
@@ -749,11 +860,20 @@ def _replaceInSequences(self, label, new):
...
+            return
+        old = getattr(self,label)
+        for task in self.schedule_()._orderedTasks:
+            task.replace(old, new)
     def globalReplace(self,label,new):
-        """ Replace the item with label 'label' by object 'new' in the process and all sequences/paths"""
+        """ Replace the item with label 'label' by object 'new' in the process and all sequences/paths/tasks"""
         if not hasattr(self,label):
             raise LookupError("process has no item of label "+label)
-        self._replaceInSequences(label, new)
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: does `setattr` do this job now?

File: FWCore/ParameterSet/python/Config.py [hunk 8]
```diff
@@ -931,19 +1066,38 @@ def __extractPSet(self,pset):
...
+        all_modules_onTasksOrScheduled = { key:value for key, value in all_modules.iteritems() if value in nodeVisitor.modules }
+        self._insertManyInto(adaptor, "@all_modules", all_modules_onTasksOrScheduled, True)
+        # Same as nodeVisitor except this one visits all the Tasks attached
+        # to the process.
+        processNodeVisitor = NodeVisitor()
+        for pTask in self.tasks.itervalues():
+            pTask.visit(processNodeVisitor)
+        esProducersToEnable = {}
+        for esProducerName, esProducer in self.es_producers_().iteritems():
+            if esProducer in nodeVisitor.esProducers or not esProducer in processNodeVisitor.esProducers:
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: perhaps a parenthesis would help

**Review by Dr15Jones — CHANGES_REQUESTED** [review]
I'll look at this more after dinner.

File: Configuration/Applications/python/ConfigBuilder.py [hunk 9]
```diff
@@ -1665,10 +1657,11 @@ def prepare_RECOBEFMIX(self, sequence = "reconstruction"):
...
     def prepare_PAT(self, sequence = "miniAOD"):
         ''' Enrich the schedule with PAT '''
-	self.prepare_PATFILTER(self)
-        self.loadDefaultOrSpecifiedCFF(sequence,self.PATDefaultCFF,1) #this is unscheduled
-	if not self._options.runUnscheduled:	
-		raise Exception("MiniAOD production can only run in unscheduled mode, please run cmsDriver with --runUnscheduled")
+        self.prepare_PATFILTER(self)
+        self.loadDefaultOrSpecifiedCFF(sequence,self.PATDefaultCFF)
+        self.scheduleSequence(sequence.split('.')[-1], 'pat_step')
+        if not self._options.runUnscheduled:
```
Comment (wddgit) [inline Configuration/Applications/python/ConfigBuilder.py]: runUnscheduled is in the interface of cmsDriver as an option. I was hesitant to change its name because it is in that interface. You are right that after this PR it controls whether the convertToUnscheduled function is called or not and also the call to
"self.pythonCfgCode += self.addCustomise(1)". This brings in some other special customisation calls through the cmsDriver option --customise_unsch. I was also hesitant to change that because it is in the cmsDriver interface. I did not check to see if anything uses that cmsDriver option. I do not know what PAT does if runUnscheduled is not set, but I left the check because it was there before. My first thought is that this is involved enough that if we want to change this, it might be better in a separate PR. In an early version of my code I had actually removed the runUnscheduled variable from ConfigBuilder.py, but then I put it back when I saw those things in the cmsDriver interface. But I will include changes to this in this PR if you ask me to. I agree that the cmsDriver interface could be improved in light of the changes that are happening.

Comment (Dr15Jones) [inline Configuration/Applications/python/ConfigBuilder.py]: If the original use of `runUnscheduled` also ran the `convertToUnscheduled` then I am fine with the present use.

Comment (wddgit) [inline Configuration/Applications/python/ConfigBuilder.py]: It did.

File: Configuration/Applications/python/ConfigBuilder.py [hunk 10]
```diff
@@ -2160,6 +2153,11 @@ def prepare(self, doChecking = False):
 
 	self.pythonCfgCode += result
 
+        from PhysicsTools.PatAlgos.tools.helpers import schedulePatAlgosEndPath
+        schedulePatAlgosEndPath(self.process)
+        self.pythonCfgCode+="from PhysicsTools.PatAlgos.tools.helpers import schedulePatAlgosEndPath\n"
+        self.pythonCfgCode+="schedulePatAlgosEndPath(process)\n"
+
```
Comment (wddgit) [inline Configuration/Applications/python/ConfigBuilder.py]: No, it did not have this end path before. I had to associate the task with some path, end path or the top level schedule that was guaranteed to exist. I did not find such a thing so I created one. That function in helpers.py will create one task and one end path if it has not already been created and then attach it to the schedule. I am open to suggestions as to a better way to do this, but this does work. The "tools" in PAT often create EDProducers and attach them to the process and there  is often no obvious task to attach the EDProducer to or sequence to attach the task to. Usually what is available in the tool function is the process.

File: Configuration/Applications/python/Options.py [hunk 11]
```diff
@@ -99,7 +99,7 @@
                   dest="runsScenarioForMC")
 
 parser.add_option("--runUnscheduled",
-                  help="Enable unscheduled mode",
+                  help="Automatically convert configuration to run unscheduled the EDProducers/EDFilters that were scheduled",
```
Comment (wddgit) [inline Configuration/Applications/python/Options.py]: Yes. Before it set the allowUnscheduled flag, did the conversion, and called the special customise functions. The difference is only that now it does not set the allowUnscheduled flag in the options ParameterSet.

File: FWCore/Framework/src/StreamSchedule.cc (hunk 1 above)
Comment (wddgit) [inline FWCore/Framework/src/StreamSchedule.cc]: Actually it is good because at this point it still includes EDAnalyzers and OutputModules in addition to EDProducers and EDFilters. Inside workerManager_.addToUnscheduledWorkers it splits the list into unscheduledLabels and shouldBeUsedLabels.

File: FWCore/ParameterSet/python/SequenceTypes.py [hunk 12]
```diff
@@ -830,13 +1228,192 @@ def __call__(self,test):
...
+            else:
+                taskContents.add(i.dumpSequencePython(options))
+        s=''
+        iFirst = True
+        for item in sorted(taskContents):
+            if not iFirst:
+                s += ", "
+            iFirst = False
+            s += item
+        return "cms.Task(" + s + ")"
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/SequenceTypes.py]: I don't think this will work if there are more than 255 entries.

File: FWCore/ParameterSet/python/SequenceTypes.py [hunk 13]
```diff
@@ -830,13 +1228,192 @@ def __call__(self,test):
...
+           Returns 'True' if a replacement occurs."""
+        # This works for either argument being of type EDProducer, EDFilter, ESProducer,
+        # ESSource, Service, or Task.
+        #
+        # Mostly this is very intuitive, but there are some complications in cases
+        # where objects that contain other objects are involved. See the comments
+        # for the _MutatingSequenceVisitor.
+
+        if not original._isTaskComponent() or (not replacement is None and not replacement._isTaskComponent()):
+           raise TypeError("replace only works with objects that can be placed on a Task")
```
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/SequenceTypes.py]: Could be useful to also print out the type that was being used as a replacement.

File: FWCore/ParameterSet/python/Config.py (hunk 3 above)
Comment (wddgit) [inline FWCore/ParameterSet/python/Config.py]: When __delattr__ is called from anywhere other than __setattr__, the new code does everything it did before and some new additional things. The additional things are that it removes deleted modules from sequences, paths, endpaths and tasks. This will help with the problem that had been reported before (in #15792) of None's showing up in the output of dumpPython in configurations where a module was deleted. It also sets a flag in a service so the service knows it is no longer attached to the process.

Previously, __setattr__ called __delattr__, but now instead it calls a new function which does almost exactly the same things except it does not remove the module from sequences, paths, end paths and tasks. The new function is called _delattrFromSetattr. Both this new function and __delattr__ call a new function _delHelper which contains the common parts.

File: FWCore/PythonParameterSet/python/testTask_cff.py [hunk 14]
```diff
@@ -0,0 +1,248 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (Dr15Jones) [inline FWCore/PythonParameterSet/python/testTask_cff.py]: Don't we normally put such test configs into the /test directory?

File: Configuration/Applications/python/ConfigBuilder.py (hunk 9 above)
Comment (Dr15Jones) [inline Configuration/Applications/python/ConfigBuilder.py]: OK

File: Configuration/Applications/python/ConfigBuilder.py (hunk 10 above)
Comment (Dr15Jones) [inline Configuration/Applications/python/ConfigBuilder.py]: @slava77 how would you like to handle this?

File: Configuration/Applications/python/Options.py (hunk 11 above)
Comment (Dr15Jones) [inline Configuration/Applications/python/Options.py]: OK

File: FWCore/Framework/src/StreamSchedule.cc (hunk 1 above)
Comment (Dr15Jones) [inline FWCore/Framework/src/StreamSchedule.cc]: OK

File: FWCore/ParameterSet/python/Config.py (hunk 3 above)
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: OK

File: PhysicsTools/PatAlgos/python/tools/jetTools.py [hunk 15]
```diff
@@ -86,17 +90,27 @@ def setupJetCorrections(process, knownModules, jetCorrections, jetSource, pvSour
...
         process.load( "JetMETCorrections.Configuration.JetCorrectorsAllAlgos_cff") # FIXME: This adds a lot of garbage
-
+        # I second the FIXME comment on the last line. When I counted it, this brought in 344 EDProducers
+        # to be available to run unscheduled. All jet correctors, probably some small fraction of which
+        # are actually used.
+        patAlgosToolsTask.add(process.jetCorrectorsAllAlgosTask)
         _payloadType = jetCorrections[0].split(_type)[0].lower()+_type
         if "PF" in _type :
             setattr(process, jetCorrections[0]+'L1FastJet', getattr(process, _payloadType+'L1FastjetCorrector').clone(srcRho=cms.InputTag('fixedGridRhoFastjetAll')))
+            patAlgosToolsTask.add(getattr(process, jetCorrections[0]+'L1FastJet'))
```
Comment (wddgit) [inline PhysicsTools/PatAlgos/python/tools/jetTools.py]: Yes. I think you are right. That would be better. I will change that. It will probably be Monday before I update the PR.

File: FWCore/ParameterSet/python/Config.py (hunk 4 above)
Comment (wddgit) [inline FWCore/ParameterSet/python/Config.py]: PathValidator and a few other similar classes check that the elements in a Path are attributes of the process. They usually do this by checking that the module has a label. Now one can put a Service on a Task, but services do not have labels. So to perform the same check I added a new variable that indicates whether a service is attached to the process. Its name is '_inProcess'. The particular line above is setting this false when a service that was attached to the process is being replaced by a different instance of that same service.

File: FWCore/ParameterSet/python/Config.py (hunk 5 above)
Comment (wddgit) [inline FWCore/ParameterSet/python/Config.py]: Probably. I will try.

File: Configuration/Applications/python/ConfigBuilder.py (hunk 10 above)
Comment (slava77) [inline Configuration/Applications/python/ConfigBuilder.py]: @Dr15Jones please clarify what is here to be handled?

Comment (Dr15Jones) [inline Configuration/Applications/python/ConfigBuilder.py]: @slava77 do you find it acceptable to add a dummy EndPath to the PAT purely to drag along the needed Tasks?
@wddgit What alternative implements have you considered and what where their strengths and weaknesses?

File: FWCore/ParameterSet/python/Config.py (hunk 6 above)
Comment (wddgit) [inline FWCore/ParameterSet/python/Config.py]: Yes. Just to be sure I ran a test and it works. If I dump a process with more than 255 paths that works and I can also execute the configuration created by dumpPython in that case.

File: FWCore/ParameterSet/python/Config.py (hunk 7 above)
Comment (wddgit) [inline FWCore/ParameterSet/python/Config.py]: Yes. Before it was calling _replaceInSequences twice, once directly and once in setattr.

File: FWCore/ParameterSet/python/Config.py (hunk 8 above)
Comment (wddgit) [inline FWCore/ParameterSet/python/Config.py]: I will add parenthesis here and in the 2 other similar lines. Although it does not affect how it executes, I agree it makes the code easier to read.

File: FWCore/ParameterSet/python/SequenceTypes.py [hunk 16]
```diff
@@ -187,19 +189,36 @@ class _ModuleSequenceType(_ConfigureComponent, _Labelable):
...
-        if len(arg)==1:
-            _checkIfSequenceable(self, arg[0])
+        tasks = arg
+        if len(arg) > 0 and isinstance(arg[0], _Sequenceable):
             self._seq = _SequenceCollection()
             arg[0]._appendToCollection(self._seq._collection)
+            tasks = arg[1:]
         self._isModified = False
+        self._tasks = set()
+        self._orderedTasks = list()
```
Comment (wddgit) [inline FWCore/ParameterSet/python/SequenceTypes.py]: In the PR as currently written, the elements of a Task, and the tasks in Paths, EndPaths, Sequences and Schedules are held in a set to prevent duplicate entries and also held in a list that is used for almost everything else. The main purpose of also having a list is to keep the iteration order fixed and reproducible. So for example if a user runs a visitor over a Path, that user would see the Tasks and contents of the Task always in the same order. That order would not depend on the order of iteration over a Python set which is not fixed and could vary. In a lot of cases this variation in order would not matter. In principle cmsRun will not be affected in any way by the order of Tasks or the order of the items on a Task. But that is not actually true when one considers the complicated ways the various CMS defined python functions that replace modules and modify configurations work. One can imagine pathological cases where the order does matter when various replace and delete and expansion operations are going on. And even something simple like printing out a text file from a Python visitor and trying to use it as a reference and "diff" against it might fail if the visitation order depended on the iteration order of Python sets.

File: FWCore/ParameterSet/python/SequenceTypes.py [hunk 17]
```diff
@@ -219,31 +238,46 @@ def __iadd__(self,rhs):
...
-    def dumpSequencePython(self):
+            s =self._seq.dumpSequencePython(options)
+        associationContents = set()
+        for task in self._orderedTasks:
+            if task.hasLabel_():
+                associationContents.add(_Labelable.dumpSequencePython(task, options))
+            else:
+                associationContents.add(task.dumpPythonNoNewline(options))
+        for iString in sorted(associationContents):
+            if s:
```
Comment (wddgit) [inline FWCore/ParameterSet/python/SequenceTypes.py]: This one was a bug. Thanks for catching it. I will fix it.

File: FWCore/ParameterSet/python/SequenceTypes.py (hunk 12 above)
Comment (wddgit) [inline FWCore/ParameterSet/python/SequenceTypes.py]: I will fix this one also.

File: FWCore/ParameterSet/python/Config.py (hunk 4 above)
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: OK

File: FWCore/ParameterSet/python/Config.py (hunk 6 above)
Comment (Dr15Jones) [inline FWCore/ParameterSet/python/Config.py]: OK

File: FWCore/ParameterSet/python/SequenceTypes.py [hunk 18]
```diff
@@ -488,34 +561,97 @@ def copy(self):
...
+            else:
+                try:
+                    # Call this just to check that taskList is a list or other sequenceable object
+                    self.dummy(*taskList)
+                except:
+                    raise RuntimeError("The Schedule constructor argument with keyword 'tasks' must have a\nlist (or other iterable object) as its value")
+                if taskList:
+                    self.associate(*taskList)
+
+    def dummy(self, *args):
```
Comment (wddgit) [inline FWCore/ParameterSet/python/SequenceTypes.py]: I will make this change.

File: FWCore/ParameterSet/python/SequenceTypes.py (hunk 13 above)
Comment (wddgit) [inline FWCore/ParameterSet/python/SequenceTypes.py]: I will add this extra info to the exception message.

File: FWCore/PythonParameterSet/python/testTask_cff.py (hunk 14 above)
Comment (wddgit) [inline FWCore/PythonParameterSet/python/testTask_cff.py]: As we discussed I will move it to python/test. It still needs to be under python because one of its purposes is to test the behavior of the Process load function and python files in the test directory cannot be loaded normally by design of the way it looks for python files to import.

**Comment by slava77** [comment]
should this become a 90X feature?

**Comment by wddgit** [comment]
Closing. I will reopen this with a new PR on 9_0_X with commits that include the fixes/improvements related to comments to this PR, reference this PR and rebase it.

**Comment by Dr15Jones** [comment]
Yes, this should be 9_0_X.
